[core][a11y] Enable Escape key dismissal for Tooltips

ggdouglas opened the pull request:
Fixes #7594

## Summary

Adds Escape key support to dismiss Tooltips, bringing Blueprint into compliance with **WCAG 2.1 Success Criterion 1.4.13** and the **WAI-ARIA Tooltip Pattern**.

| Before | After |
|--------|--------|
| ![before](https://github.com/user-attachments/assets/858be233-407a-4b86-848f-e8c1ce85434d) | ![after](https://github.com/user-attachments/assets/fed5dc90-c148-40cf-8f8e-86a022f990d0) | 

## Problem

Tooltips in Blueprint could not be dismissed using the Escape key, which violates:

1. **[WCAG 2.1 SC 1.4.13 (Content on Hover or Focus)](https://www.w3.org/WAI/WCAG21/Understanding/content-on-hover-or-focus.html)** - requires a mechanism to dismiss content without moving pointer hover or keyboard focus
2. **[WAI-ARIA Tooltip Pattern](https://www.w3.org/WAI/ARIA/apg/patterns/tooltip/)** - explicitly specifies that pressing Escape should dismiss tooltips

### Previous Behavior
- Tooltip had `canEscapeKeyClose={false}` hardcoded
- Even if enabled, Escape wouldn't work because Overlay2's handler only fires when the container has focus
- Tooltips use `autoFocus={false}` to avoid stealing focus on hover

## Solution

Added a document-level Escape key listener to Overlay2 that activates when `autoFocus={false}`. Tooltip no longer hardcodes `canEscapeKeyClose={false}`, allowing it to inherit the default `true` value from Overlay2 while keeping `autoFocus={false}` to prevent focus steal on hover.

### Testing

1. Hover over a tooltip target
2. Tooltip appears
3. Press `Escape` key
4. Tooltip dismisses without moving pointer
5. Focus remains on the original element

ggdouglas committed:
Add functionality to allow Tooltips to be esc-key closed

svc-palantir-github commented:
<h3>Add functionality to allow Tooltips to be esc-key closed</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/4bfb6b20-a97f-4834-9d26-cb20e91cd3ba/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/4bfb6b20-a97f-4834-9d26-cb20e91cd3ba/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/4bfb6b20-a97f-4834-9d26-cb20e91cd3ba/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/4bfb6b20-a97f-4834-9d26-cb20e91cd3ba/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

mm-wang commented on packages/core/test/tooltip/tooltipTests.tsx:
```diff
@@ -155,6 +155,79 @@ describe("<Tooltip>", () => {
         });
     });
 
+    describe("keyboard interactions", () => {
```
**question**: Open to using React Testing Library for new tests?

ggdouglas replied:
Sure! There aren't that many tests here in this suite, so I've opted to replace them all with RTL for consistency. 56a9874525c19104b0dd726042770c7099c5d9ca

mm-wang reviewed (commented):
Any chance we can get a test for multiple overlays?

ggdouglas committed:
Rewrite Tooltip tests with RTL

svc-palantir-github commented:
<h3>Rewrite Tooltip tests with RTL</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/68ff961f-569f-4fa2-a6e3-3868d108b40f/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/68ff961f-569f-4fa2-a6e3-3868d108b40f/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/68ff961f-569f-4fa2-a6e3-3868d108b40f/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/68ff961f-569f-4fa2-a6e3-3868d108b40f/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ggdouglas committed:
Write test to verify closing multiple Tooltips with Esc

svc-palantir-github commented:
<h3>Write test to verify closing multiple Tooltips with Esc</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/1ee1660b-0025-4400-ad3b-de2f128e679e/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/1ee1660b-0025-4400-ad3b-de2f128e679e/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/1ee1660b-0025-4400-ad3b-de2f128e679e/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/1ee1660b-0025-4400-ad3b-de2f128e679e/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

ggdouglas committed:
Merge branch 'develop' into gdouglas/tooltip-a11y-esc

svc-palantir-github commented:
<h3>Merge branch 'develop' into gdouglas/tooltip-a11y-esc</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/daa04851-eec9-44b9-b538-051903188c0b/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/daa04851-eec9-44b9-b538-051903188c0b/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/daa04851-eec9-44b9-b538-051903188c0b/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/daa04851-eec9-44b9-b538-051903188c0b/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

mm-wang reviewed (commented):
While I don't see an issue with the code, when I try the Tooltip example in the demo app, I can't escape the tooltip. Even when I'm checking out the branch and isolating the example. Is it working for you?

ggdouglas commented:
> While I don't see an issue with the code, when I try the Tooltip example in the demo app, I can't escape the tooltip. Even when I'm checking out the branch and isolating the example. Is it working for you?

@mm-wang This behavior exists in the demo due to the forced open Popover in the PopoverExample, which steals focus away into the Popover's overlay focus trap: https://github.com/palantir/blueprint/blob/9c1e4cd71973dede59ff3f8329e76a0619cf37e6/packages/demo-app/src/examples/PopoverExample.tsx#L47-L59

We should maybe consider adding `autoFocus={false}` on that example to prevent it from interfering with other examples on the page, but perhaps it would be good to leave it as is as an example of how these interactions currently behave in Blueprint. 🤔 In any case, I think this should be a pretty uncommon edge case when compared to dismissing tooltips in general.

ggdouglas committed:
Handle native KeyboardEvent in handleOverlayClose

handleOverlayClose now correctly detects both native KeyboardEvents
and React synthetic keyboard events. This fixes a case where Tooltips
were still not Escape-key closing when the active element was focused
on the Tooltip's target.

svc-palantir-github commented:
<h3>Handle native KeyboardEvent in handleOverlayClose</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/ec554964-c688-40d9-b037-12d2f38ed6b2/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/ec554964-c688-40d9-b037-12d2f38ed6b2/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/ec554964-c688-40d9-b037-12d2f38ed6b2/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/ec554964-c688-40d9-b037-12d2f38ed6b2/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>

mm-wang commented:
I'm attaching a screen recording where I've pulled your branch with the latest change and have isolated the tooltip example. I don't see escape closing the tooltip. Is there something that I am missing?

https://github.com/user-attachments/assets/5bab41b0-2009-477a-9640-89a841c874bc

mm-wang reviewed (approved):
LGTM! Sorry for the mishap in testing!
